### ISSUE TITLE
sequential read speed vs mfsreadaheadleng limit

### ISSUE THREAD

**Issue by oszafraniec** [moosefs/moosefs#310]
**Case:** we are hosting files on MFS that are downloaded one by one via FTP. After upgrading connection from 100Mbps to 1Gbps we encountered a problem, that single download can perform max @~300Mbps. After some digging we found this: https://sourceforge.net/p/moosefs/mailman/message/33792353/ and changed mfsreadaheadleng from default 1MB to 2MB and speed increased to ~500Mbps. @xandrus suggested at SourceForge that we can rise this to 4MB or more, but... mfsreadaheadleng can be increased max to 2MB, based on the manual and in practice as well (we can set it up to 4MB, but it gives no speed improvement).

**Request:** please, make mfsreadaheadleng (and any other options related to it) not limited to 2MB... so we can saturate 1Gbps connection in single thread. I can't find this hard-limit in the code to make a pull request ;) 

We use MFS 3.0.103 or higher, but I think it's in every version.

**Comment by oszafraniec** [comment]
I think I know why I can't find this in the code... This relates to PRO version only :| Mounting to "community" version works. Mounting to PRO version gives this:

error in fuse_mount
read ahead length too big (4194304 B) - decresed to 2 MiB

:(

**Comment by acid-maker** [comment]
Hmm. With default settings we were able to saturate 100Gb (using libmfsio and many threads of course). The readaheadleng is actually multiplied by a factor that is increased when subsequent calls are sequential and decreased when subsequent calls are more random. Maximum value of this factor is 64, so when you read a file from MooseFS with readaheadleng set to 2MB you actually read data using 128MB blocks (yes - two chunks at the same time) - increasing this value rather doesn't make sense.

Likely the reason you got better results with 2MB is beacuse the system read two chunks at the same time, but still the read time per chunks is almost the same and quite slow.

Maybe even increasing this to 4MB will cause four chunks to be read at the same time and maybe it could help you a bit, but in my opinion this is not the solution to your problem. It is likely that your problem lies elsewhere and is network/harddrives related.

**Comment by oszafraniec** [comment]
> and many threads of course
@acid-maker , of course we can saturate our 1Gb connection with many threads, not a problem. So this is not network/bandwidth related or harddrives, as we can do it with "many threads of course". The problem we have is with 1 thread.

I can agree that this can be somehow related to network latance, as @xandrus described years ago. As we have clusted on heavy load, small operations are fine (many clients/many files uploaded), but at the same time single threaded file download that I described dropted below 10Mb. Yes, the network is on higher load during but at the same time small files fly and saturate more than 1Gb easly. Master's CPU is 50%, there is still network bandwith avaible and chunkservers have max 8-10 load. Everything works fine besides 1-threaded-big-file-downloads.

And if you have 10Gb in your lab, please make some one-big-file like 100GB and try to saturate your 10Gb uplink while downloading it. With one thread. Just copy it back to the client and see how fast it will go (best to RAM drive so local drive on a client is not a bottleneck). And let me know your result.

**Comment by acid-maker** [comment]
I don't think that network latency is so important when we are reading huge sequential files - low latency is rather needed for reading small files or reading one file randomly.

Maybe you saturated your hard drives?

Our current results on 10Gb network (5 chunkservers, 3 mechanical hard disks each - so 15 mechanical disks). Tested using FIO.

Default `readaheadleng`
- Sequential read using 128k block: 348MiB/s
- Sequential read using 1M block: 420MiB/s

`readaheadleng` set to 2097152:
- Sequentila read using 1M block: 392MiB/s

As you can see we are able to utilize about 1/3 of this network with only one client and one thread. Very likely the bottle neck here is the speed of our drives (about 100MiB/s).

**Comment by oszafraniec** [comment]
Can you provide full fio command line that you are using for test? I will check it against our cluster to compare.

**Comment by xandrus** [comment]
Hi,
This is our command line:
```
fio --directory=/moosefs/fio --bs=128k --numjobs=1 --name=write-test-job --rw=read --size=100G --group_reporting
```

**Comment by xandrus** [comment]
Hi,
Just in case we have made a test using dd and this is the result:
```
dd if=/moosefs/fio/write-test-job.0.0 of=/dev/null bs=128k
819200+0 records in
819200+0 records out
107374182400 bytes (107 GB, 100 GiB) copied, 262.518 s, 409 MB/s
```

**Comment by oszafraniec** [comment]
Thank you very much for the feedback. It can be that we are wrong and we where making mistake thinking that MFS is the bottleneck (@xandrus post directed us there, sorry). We made some tests like you did and we have strange results while testing from different network locations (from 180MB/s to 50MB/s). We will investigate our network and share later what we will find.

**Comment by inkdot7** [comment]
If I understand the responses of @acid-maker correctly, the MFS client will only have two outstanding read requests at the same time?

I.e. it can for single-threaded reading only achieve 2 x the transfer bandwidth of the underlying storage in the chunk servers, which in the case of modern hard drives is on the order of 250 MB/s, and perhaps 100 MB/s in the 4 TB max hard-drive size era.

In order to saturate a 10 Gbps link with 100 MB/s drives would require the MFS client to have at least 10, perhaps 15 read requests in-flight at the same time.  While `mfsreadaheadsize` can be set to 2048 MB, one would also need to be able to set  `mfsreadaheadleng` to 16 MB, such that it together with the internal multiplication of 64 reaches 16 chunks being read simultaneously?

Are there other internal assumptions that require the number of outstanding read requests be limited to 2?

(Btw, thanks for a great filesystem and easy interface - it has been a pleasure to set up for test!  Am using 3.0.115 (Debian Bullseye packages.))

**Comment by chogata** [comment]
In general, we don't want to be too greedy. Read ahead of 16 chunks, when the real read request is for, let's say, 200 bytes, makes no sense. The client assumes it's not the only process on a machine, there could be other MooseFS clients mounted, other filesystems. They also need their RAM for cache and other stuff. Plus, there would be other clients on other machines, that also want to use the network and chunkservers' resources. It would be theoretically possible to make a client that exhausts RAM for cache, but we never wanted to actually make it so. Those are, in fact, our "internal assumptions" :) - that MooseFS is used as multi-user, multi machine system and one machine shouldn't exhaust all the resources available.

**Comment by VideoFX** [comment]
I also landed here after noticing I was having a hard time getting full speeds on a single download of a large file. I happened to notice that lowering "mfsreadaheadtrigger" to values like 1048576 or 512000 helped to get speeds up significantly, however, I am not sure why that is working. It would be nice to let 'er rip in a single thread or single user scenario. Brilliant work, thank you for letting me (us) post and test here.

**Comment by inkdot7** [comment]
For my use-case it would indeed be interesting to use much more of the available disk bandwidths for one or a few readers.

To just try this out, I made some small changes in the client to activate and keep more than two chunks of readahead active at the same time (and allow a larger `mfsreadaheadleng` value):

https://github.com/inkdot7/moosefs/tree/readahead_many

(Note: the changes are a very quick-and-dirty hack, just to see if it is possible.  Nice thing is that no other changes were necessary, i.e. only the client readahead was modified.)

Tests on a system with a total of 26 HDDs (~4 TB, i.e. old, ~100 MB/s per drive) in 3 machines, and a total of 8 SSDs (again, rather old) in 3 machines and 4 NVMEs in 2 machines.  Storage type selected by storage classes.  Networking 10 Gbps.  The client is on one of the machines with NVME drives.

| mfsmount                | HDD          | SSD          | NVME         | Notes |
| ----------------------- | ------------ | ------------ | ------------ | ------------ |
| plain, no opts.         | 260 MB/s     | 563 MB/s     | 1315 MB/s    |
| plain, 2048, 2M, 1024   | 277 MB/s     | 617 MB/s     | 1433 MB/s    |
| plain, 2048, - , 1024   | 293 MB/s     | 409 MB/s     | 1442 MB/s    |
| plain, -, -, 1024       | 264 MB/s     | 559 MB/s     | 1364 MB/s    |
| plain, no opts (again)  | 257 MB/s     | 548 MB/s     | 1203 MB/s    |
| **hacked**                  | **593 MB/s**     | **946 MB/s**     | **1442 MB/s**    |  <- Higher throughput |
| hacked, 2048, 32M ,1024 | 642 MB/s     | 1068 MB/s    | 1664 MB/s    |

The three numbers in the mount column are the `mfsmount` options `mfsreadaheadsize`, `mfsreadaheadleng`, and `mfsreadaheadtrigger`.  Their effect is rather small.

(Note, the quick hack above does not respect the value of the `mfsreadaheadleng` option, explaining the higher throughput even without options.)

It seems that even if user-space requests a large amount of data in one `read()` call, all the data is not retrieved by the mfsclient in parallel, but the the read-ahead logic seems to be needed/used also in that case?

Is the explanation that the FUSE layer only gets the reads in 128 kiB pieces, and thus needs to use the read-ahead logic, even for the case when user-space explicitly has asked for a large amount of data at once?

The tests were done for about 70 GB of data each with `dd`, where it was told to read in pieces of large buffers, 1 GB (`dd if=inputfile of=/dev/null bs=1G`).

I wonder if a new option really would be needed?  If the client respects the `mfsreadaheadleng` option, it would with the current default value only do read-ahead for 1 or 2 chunks, i.e. as before.  Only if that option would be set to a large value (than currently possible), then the client would do more aggressive read-ahead.
